File: common/src/main/java/DahRealPanda/plugins/colonyweb/texture/TextureService.java
package DahRealPanda.plugins.colonyweb.texture;

import DahRealPanda.plugins.colonyweb.util.Text;
import com.mojang.logging.LogUtils;
import net.minecraft.resources.ResourceLocation;
import org.slf4j.Logger;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * Converts an {@code itemKey} into PNG icon bytes, resolving item/block models, DO materials,
 * modded jar textures and cached vanilla assets, with an in-memory + disk cache and a
 * generated placeholder fallback.
 *
 * <p>Blocks get the icon a player sees in their inventory: the model's own geometry rendered
 * isometrically under the vanilla GUI transform. Items whose model is a flat sprite (wheat, a
 * stick, a door) keep that sprite, because that is equally what the inventory shows.</p>
 *
 * <p>Domum Ornamentum blocks go one step further and have each material component's texture
 * substituted into the model first, so a "Brick Extra Shingle" looks like a shingle made of
 * brick rather than a flat brick square.</p>
 */
public final class TextureService {
    private static final Logger LOGGER = LogUtils.getLogger();

    /** Edge length of generated 3D block icons. */
    private static final int ICON_SIZE = 64;

    /** Texture-variable / component-id words that carry no matching signal. */
    private static final Set<String> GENERIC_WORDS = Set.of(
            "domum", "ornamentum", "minecraft", "material", "materials", "block", "blocks",
            "texture", "textures", "the", "all", "any", "default", "main");

    private final VanillaAssetProvider vanilla;
    private final ModelResolver modelResolver;
    private final PngCache cache;

    public TextureService(Path dataDir, VanillaAssetProvider vanilla) {
        this.vanilla = vanilla;
        this.modelResolver = new ModelResolver(vanilla);
        this.cache = new PngCache(dataDir);
    }

    /** Get PNG bytes for a texture key (never null; returns a placeholder on miss). */
    public byte[] getPng(String itemKey) {
        byte[] cached = cache.get(itemKey);
        if (cached != null) {
            return cached;
        }
        byte[] png = resolve(itemKey);
        if (png == null) {
            png = placeholder();
        }
        cache.put(itemKey, png);
        return png;
    }

    private byte[] resolve(String itemKey) {
        String base = itemKey;
        int hashIdx = itemKey.indexOf('#');
        if (hashIdx >= 0) {
            base = itemKey.substring(0, hashIdx);
        }

        // Domum Ornamentum: render the block's real geometry in its real materials.
        if (DomumOrnamentumResolver.isDomumKey(itemKey)) {
            byte[] rendered = renderDomum(itemKey, base);
            if (rendered != null) {
                return rendered;
            }
        }

        // Fall back to a flat swatch of the primary material.
        Optional<String> material = DomumOrnamentumResolver.materialForKey(itemKey);
        if (material.isPresent()) {
            byte[] mat = pngForRegistryName(material.get());
            if (mat != null) {
                return mat;
            }
        }

        // Everything else: draw whatever the inventory slot would draw. Items whose model is a
        // flat sprite report no geometry and drop through to their texture unchanged.
        byte[] rendered = renderInventoryIcon(base);
        if (rendered != null) {
            return rendered;
        }

        return pngForRegistryName(base);
    }

    /**
     * Render a block the way a player sees it in their inventory.
     *
     * <p>A block's inventory icon is its model drawn under the vanilla GUI transform, not one of
     * its face textures — a furnace shows its front, a stair shows its steps, a fence shows its
     * inventory post. Picking a single texture out of the model instead (which is all
     * {@link ModelResolver#resolveTexture} can do) turns every one of those into a flat swatch
     * of the wrong thing.</p>
     *
     * <p>Beds and shields are asked about first: their models are deliberately empty because the
     * game draws them from Java, so only {@link BuiltinEntityModels} has geometry for them.</p>
     *
     * @return null when the item has no geometry to draw, so the caller keeps the flat texture
     */
    private byte[] renderInventoryIcon(String registryName) {
        try {
            ResourceLocation rl = ResourceLocation.tryParse(registryName);
            if (rl == null) {
                return null;
            }
            Optional<BlockModel> resolved = BuiltinEntityModels.forItem(rl.getNamespace(), rl.getPath())
                    .or(() -> modelResolver.resolveInventoryModel(rl.getNamespace(), rl.getPath()));
            if (resolved.isEmpty()) {
                return null;
            }
            BlockModel model = resolved.get();
            BufferedImage icon = IsometricRenderer.render(
                    model, ref -> imageForTextureRef(model.resolveTextureRef(ref)), ICON_SIZE);
            return icon != null ? encode(icon) : null;
        } catch (Throwable t) {
            LOGGER.debug("[ColonyWeb] inventory render failed for {}", registryName, t);
            return null;
        }
    }

    // ------------------------------------------------------------------
    // Domum Ornamentum 3D icons
    // ------------------------------------------------------------------

    /**
     * Render a DO variant isometrically. Returns null (so the caller falls back to a flat
     * texture) when the model has no parseable geometry or no material resolves.
     */
    private byte[] renderDomum(String itemKey, String baseRegistryName) {
        try {
            Map<String, String> components = DomumOrnamentumResolver.componentsForKey(itemKey);
            if (components.isEmpty()) {
                return null;
            }
            ResourceLocation rl = ResourceLocation.tryParse(baseRegistryName);
            if (rl == null) {
                return null;
            }
            Optional<BlockModel> model = modelResolver.resolveModel(rl.getNamespace(), rl.getPath());
            if (model.isEmpty()) {
                return null;
            }
            Map<String, BufferedImage> faces = assignMaterials(model.get(), components);
            if (faces.isEmpty()) {
                return null;
            }
            BufferedImage icon = IsometricRenderer.render(model.get(), faces::get, ICON_SIZE);
            return icon != null ? encode(icon) : null;
        } catch (Throwable t) {
            LOGGER.debug("[ColonyWeb] 3D render failed for {}", itemKey, t);
            return null;
        }
    }

    /**
     * Decide which material texture each of the model's texture variables should use.
     *
     * <p>DO does the same substitution at render time, but its component-to-variable wiring
     * is not described in the model JSON. Variables are therefore matched to components by
     * name first, then positionally in the block's declared component order — so the shape is
     * always right, and multi-material blocks land their materials in a stable order.</p>
     */
    private Map<String, BufferedImage> assignMaterials(BlockModel model, Map<String, String> components) {
        List<String> refs = model.usedTextureRefs();
        if (refs.isEmpty()) {
            return Map.of();
        }
        List<String> componentIds = new ArrayList<>(components.keySet());
        List<BufferedImage> componentImages = new ArrayList<>();
        for (String id : componentIds) {
            componentImages.add(imageForRegistryName(components.get(id)));
        }

        Map<String, BufferedImage> assigned = new LinkedHashMap<>();
        boolean[] taken = new boolean[componentIds.size()];

        // 1. Match by name — a "#shingle" variable should take the shingle component.
        for (String ref : refs) {
            String varName = ref.startsWith("#") ? ref.substring(1) : Text.pathOf(ref);
            for (int i = 0; i < componentIds.size(); i++) {
                if (!taken[i] && sharesWord(varName, componentIds.get(i))) {
                    assigned.put(ref, componentImages.get(i));
                    taken[i] = true;
                    break;
                }
            }
        }

        // 2. Fill the rest positionally, reusing the first material once components run out.
        int next = 0;
        BufferedImage fallback = componentImages.stream().filter(java.util.Objects::nonNull).findFirst().orElse(null);
        for (String ref : refs) {
            if (assigned.containsKey(ref)) {
                continue;
            }
            while (next < taken.length && taken[next]) {
                next++;
            }
            if (next < taken.length) {
                assigned.put(ref, componentImages.get(next));
                taken[next] = true;
            } else {
                assigned.put(ref, fallback);
            }
        }

        // 3. Anything still without pixels keeps the model's own declared texture.
        boolean any = false;
        for (String ref : refs) {
            BufferedImage image = assigned.get(ref);
            if (image == null) {
                image = imageForTextureRef(model.resolveTextureRef(ref));
                assigned.put(ref, image);
            }
            any |= image != null;
        }
        return any ? assigned : Map.of();
    }

    /** True when two identifiers share a meaningful word (case/separator insensitive). */
    private static boolean sharesWord(String a, String b) {
        Set<String> wordsA = words(a);
        wordsA.retainAll(words(b));
        return !wordsA.isEmpty();
    }

    private static Set<String> words(String raw) {
        Set<String> out = new HashSet<>();
        if (raw == null) {
            return out;
        }
        for (String part : raw.toLowerCase(Locale.ROOT).split("[^a-z0-9]+")) {
            if (part.length() >= 3 && !GENERIC_WORDS.contains(part)) {
                out.add(part);
            }
        }
        return out;
    }

    // ------------------------------------------------------------------
    // Texture loading
    // ------------------------------------------------------------------

    private final Map<String, Optional<BufferedImage>> imageCache = new HashMap<>();

    /** Decoded texture for a block/item registry name, memoized per service instance. */
    private synchronized BufferedImage imageForRegistryName(String registryName) {
        return imageCache.computeIfAbsent("r:" + registryName, key -> {
            byte[] png = pngForRegistryName(registryName);
            return Optional.ofNullable(decode(png));
        }).orElse(null);
    }

    /** Decoded texture for a direct {@code namespace:path} texture reference. */
    private synchronized BufferedImage imageForTextureRef(String textureRef) {
        if (textureRef == null) {
            return null;
        }
        return imageCache.computeIfAbsent("t:" + textureRef, key -> {
            byte[] png = loadTexturePng(textureRef);
            return Optional.ofNullable(decode(png));
        }).orElse(null);
    }

    private byte[] pngForRegistryName(String registryName) {
        ResourceLocation rl = ResourceLocation.tryParse(registryName);
        if (rl == null) {
            return null;
        }
        Optional<String> texture = modelResolver.resolveTexture(rl.getNamespace(), rl.getPath());
        if (texture.isEmpty()) {
            return null;
        }
        return loadTexturePng(texture.get());
    }

    private byte[] loadTexturePng(String textureRef) {
        ModelResolver.ParsedRef ref = ModelResolver.ParsedRef.of(textureRef);
        String assetPath = "textures/" + ref.path() + ".png";
        byte[] bytes;
        if ("minecraft".equals(ref.namespace())) {
            bytes = vanilla.readTexture(assetPath);
        } else {
            bytes = ModelResolver.classpathBytes("assets/" + ref.namespace() + "/" + assetPath);
        }
        if (bytes == null) {
            return null;
        }
        return firstFrame(bytes);
    }

    /** Some textures are animated (tall strips); crop to the top square frame. */
    private byte[] firstFrame(byte[] pngBytes) {
        try {
            BufferedImage img = ImageIO.read(new ByteArrayInputStream(pngBytes));
            if (img == null) {
                return pngBytes;
            }
            int w = img.getWidth();
            int h = img.getHeight();
            if (h > w && w > 0 && h % w == 0) {
                byte[] frame = encode(img.getSubimage(0, 0, w, w));
                return frame != null ? frame : pngBytes;
            }
            return pngBytes;
        } catch (IOException e) {
            return pngBytes;
        }
    }

    private static BufferedImage decode(byte[] png) {
        if (png == null) {
            return null;
        }
        try {
            BufferedImage read = ImageIO.read(new ByteArrayInputStream(png));
            if (read == null) {
                return null;
            }
            // Normalize to ARGB so alpha handling in the renderer is uniform.
            BufferedImage argb = new BufferedImage(read.getWidth(), read.getHeight(), BufferedImage.TYPE_INT_ARGB);
            argb.getGraphics().drawImage(read, 0, 0, null);
            return argb;
        } catch (IOException e) {
            return null;
        }
    }

    private static byte[] encode(BufferedImage image) {
        try {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            ImageIO.write(image, "png", out);
            return out.toByteArray();
        } catch (IOException e) {
            return null;
        }
    }

    /** Magenta/black checker placeholder so the UI still lays out. */
    private byte[] placeholder() {
        int size = 16;
        BufferedImage img = new BufferedImage(size, size, BufferedImage.TYPE_INT_ARGB);
        int magenta = 0xFFF800F8;
        int black = 0xFF000000;
        for (int y = 0; y < size; y++) {
            for (int x = 0; x < size; x++) {
                boolean a = (x / 8 + y / 8) % 2 == 0;
                img.setRGB(x, y, a ? magenta : black);
            }
        }
        byte[] png = encode(img);
        return png != null ? png : new byte[0];
    }
}

File: common/src/main/kotlin/DahRealPanda/plugins/colonyweb/service/TextureService.kt
package DahRealPanda.plugins.colonyweb.service

import DahRealPanda.plugins.colonyweb.model.BlockModel
import DahRealPanda.plugins.colonyweb.model.BuiltinEntityModels
import DahRealPanda.plugins.colonyweb.provider.VanillaAssetProvider
import DahRealPanda.plugins.colonyweb.renderer.IsometricRenderer
import DahRealPanda.plugins.colonyweb.repository.PngCache
import DahRealPanda.plugins.colonyweb.util.Text
import com.mojang.logging.LogUtils
import net.minecraft.resources.ResourceLocation
import org.slf4j.Logger
import java.awt.image.BufferedImage
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.nio.file.Path
import javax.imageio.ImageIO

class TextureService(dataDir: Path, private val vanilla: VanillaAssetProvider) {
    companion object {
        private val LOGGER: Logger = LogUtils.getLogger()

        private const val ICON_SIZE = 64

        private val GENERIC_WORDS = setOf(
            "domum", "ornamentum", "minecraft", "material", "materials", "block", "blocks",
            "texture", "textures", "the", "all", "any", "default", "main"
        )
    }

    private val modelResolver = ModelResolver(vanilla)
    private val cache = PngCache(dataDir)

    fun getPng(itemKey: String): ByteArray {
        var png = cache.get(itemKey)
        if (png != null) return png
        png = resolve(itemKey)
        if (png == null) {
            png = placeholder()
        }
        cache.put(itemKey, png)
        return png
    }

    private fun resolve(itemKey: String): ByteArray? {
        val base: String
        val hashIdx = itemKey.indexOf('#')
        if (hashIdx >= 0) {
            base = itemKey.substring(0, hashIdx)
        } else {
            base = itemKey
        }

        if (DomumOrnamentumResolver.isDomumKey(itemKey)) {
            val rendered = renderDomum(itemKey, base)
            if (rendered != null) return rendered
        }

        val material = DomumOrnamentumResolver.materialForKey(itemKey)
        if (material != null) {
            val mat = pngForRegistryName(material)
            if (mat != null) return mat
        }

        val rendered = renderInventoryIcon(base)
        if (rendered != null) return rendered

        return pngForRegistryName(base)
    }

    private fun renderInventoryIcon(registryName: String): ByteArray? {
        return try {
            val rl = ResourceLocation.tryParse(registryName) ?: return null
            val resolved = BuiltinEntityModels.forItem(rl.namespace, rl.path)
                ?: modelResolver.resolveInventoryModel(rl.namespace, rl.path)
                ?: return null
            val model = resolved
            val icon = IsometricRenderer.render(
                model, { ref -> imageForTextureRef(model.resolveTextureRef(ref)) }, ICON_SIZE
            )
            if (icon != null) encode(icon) else null
        } catch (t: Throwable) {
            LOGGER.debug("[ColonyWeb] inventory render failed for {}", registryName, t)
            null
        }
    }

    private fun renderDomum(itemKey: String, baseRegistryName: String): ByteArray? {
        return try {
            val components = DomumOrnamentumResolver.componentsForKey(itemKey)
            if (components.isEmpty()) return null
            val rl = ResourceLocation.tryParse(baseRegistryName) ?: return null
            val model = modelResolver.resolveModel(rl.namespace, rl.path) ?: return null
            val faces = assignMaterials(model, components)
            if (faces.isEmpty()) return null
            val icon = IsometricRenderer.render(model, { faces[it] }, ICON_SIZE)
            if (icon != null) encode(icon) else null
        } catch (t: Throwable) {
            LOGGER.debug("[ColonyWeb] 3D render failed for {}", itemKey, t)
            null
        }
    }

    private fun assignMaterials(model: BlockModel, components: Map<String, String>): Map<String, BufferedImage> {
        val refs = model.usedTextureRefs()
        if (refs.isEmpty()) return emptyMap()
        val componentIds = ArrayList(components.keys)
        val componentImages = componentIds.map { imageForRegistryName(components[it]) }.toMutableList()

        val assigned = LinkedHashMap<String, BufferedImage?>()
        val taken = BooleanArray(componentIds.size)

        for (ref in refs) {
            val varName = if (ref.startsWith("#")) ref.substring(1) else Text.pathOf(ref)
            for (i in componentIds.indices) {
                if (!taken[i] && sharesWord(varName, componentIds[i])) {
                    assigned[ref] = componentImages[i]
                    taken[i] = true
                    break
                }
            }
        }

        var next = 0
        val fallback = componentImages.firstOrNull { it != null }
        for (ref in refs) {
            if (assigned.containsKey(ref)) continue
            while (next < taken.size && taken[next]) next++
            if (next < taken.size) {
                assigned[ref] = componentImages[next]
                taken[next] = true
            } else {
                assigned[ref] = fallback
            }
        }

        var any = false
        for (ref in refs) {
            var image = assigned[ref]
            if (image == null) {
                image = imageForTextureRef(model.resolveTextureRef(ref))
                assigned[ref] = image
            }
            any = any or (image != null)
        }
        @Suppress("UNCHECKED_CAST")
        return if (any) assigned.filterValues { it != null } as Map<String, BufferedImage> else emptyMap()
    }

    private fun sharesWord(a: String, b: String): Boolean {
        val wordsA = words(a)
        wordsA.retainAll(words(b))
        return wordsA.isNotEmpty()
    }

    private fun words(raw: String?): MutableSet<String> {
        val out = LinkedHashSet<String>()
        if (raw == null) return out
        for (part in raw.lowercase().split("[^a-z0-9]+".toRegex())) {
            if (part.length >= 3 && part !in GENERIC_WORDS) {
                out.add(part)
            }
        }
        return out
    }

    private val imageCache = HashMap<String, BufferedImage?>()

    @Synchronized
    private fun imageForRegistryName(registryName: String?): BufferedImage? {
        if (registryName == null) return null
        return imageCache.getOrPut("r:$registryName") {
            val png = pngForRegistryName(registryName)
            decode(png)
        }
    }

    @Synchronized
    private fun imageForTextureRef(textureRef: String?): BufferedImage? {
        if (textureRef == null) return null
        return imageCache.getOrPut("t:$textureRef") {
            val png = loadTexturePng(textureRef)
            decode(png)
        }
    }

    private fun pngForRegistryName(registryName: String?): ByteArray? {
        val rl = ResourceLocation.tryParse(registryName) ?: return null
        val texture = modelResolver.resolveTexture(rl.namespace, rl.path) ?: return null
        return loadTexturePng(texture)
    }

    private fun loadTexturePng(textureRef: String): ByteArray? {
        val ref = ModelResolver.ParsedRef.of(textureRef)
        val assetPath = "textures/${ref.path}.png"
        val bytes = if ("minecraft" == ref.namespace) {
            vanilla.readTexture(assetPath)
        } else {
            ModelResolver.classpathBytes("assets/${ref.namespace}/$assetPath")
        }
        if (bytes == null) return null
        return firstFrame(bytes)
    }

    private fun firstFrame(pngBytes: ByteArray): ByteArray {
        return try {
            val img = ImageIO.read(ByteArrayInputStream(pngBytes)) ?: return pngBytes
            val w = img.width
            val h = img.height
            if (h > w && w > 0 && h % w == 0) {
                val frame = encode(img.getSubimage(0, 0, w, w))
                frame ?: pngBytes
            } else {
                pngBytes
            }
        } catch (_: java.io.IOException) {
            pngBytes
        }
    }

    private fun decode(png: ByteArray?): BufferedImage? {
        if (png == null) return null
        return try {
            val read = ImageIO.read(ByteArrayInputStream(png)) ?: return null
            val argb = BufferedImage(read.width, read.height, BufferedImage.TYPE_INT_ARGB)
            argb.graphics.drawImage(read, 0, 0, null)
            argb
        } catch (_: java.io.IOException) {
            null
        }
    }

    private fun encode(image: BufferedImage): ByteArray? {
        return try {
            val out = ByteArrayOutputStream()
            ImageIO.write(image, "png", out)
            out.toByteArray()
        } catch (_: java.io.IOException) {
            null
        }
    }

    private fun placeholder(): ByteArray {
        val size = 16
        val img = BufferedImage(size, size, BufferedImage.TYPE_INT_ARGB)
        val magenta = 0xFFF800F8.toInt()
        val black = 0xFF000000.toInt()
        for (y in 0 until size) {
            for (x in 0 until size) {
                val a = (x / 8 + y / 8) % 2 == 0
                img.setRGB(x, y, if (a) magenta else black)
            }
        }
        val png = encode(img)
        return png ?: ByteArray(0)
    }
}
